File: src/test/java/org/blitzortung/android/data/ParametersTest.java
package org.blitzortung.android.data;

import com.xtremelabs.robolectric.RobolectricTestRunner;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;

@RunWith(RobolectricTestRunner.class)
public class ParametersTest {

    private Parameters parameters;

    @Before
    public void setUp() {
        parameters = new Parameters();
        parameters.setOffsetIncrement(15);
        parameters.setIntervalDuration(60);
    }

    @Test
    public void testIfRealtimeModeIsDefault() {
        assertTrue(parameters.isRealtime());
    }

    @Test
    public void testGetOffsetReturnsZeroByDefault() {
        assertThat(parameters.getIntervalOffset(), is(0));
    }

    @Test
    public void testRewindInterval() {
        assertTrue(parameters.revInterval());
        assertThat(parameters.getIntervalOffset(), is(-15));

        assertTrue(parameters.revInterval());
        assertThat(parameters.getIntervalOffset(), is(-30));

        for (int i = 0; i < 23 * 4 - 2 - 1; i++) {
            assertTrue(parameters.revInterval());
        }
        assertThat(parameters.getIntervalOffset(), is(-23 * 60 + 15));

        assertTrue(parameters.revInterval());
        assertThat(parameters.getIntervalOffset(), is(-23 * 60));

        assertFalse(parameters.revInterval());
        assertThat(parameters.getIntervalOffset(), is(-23 * 60));
    }

    @Test
    public void testRewindIntervalWithAlignment() {
        parameters.setOffsetIncrement(45);

        assertTrue(parameters.revInterval());
        assertThat(parameters.getIntervalOffset(), is(-45));

        assertTrue(parameters.revInterval());
        assertThat(parameters.getIntervalOffset(), is(-90));

        for (int i = 0; i < 23 / 3 * 4; i++) {
            assertTrue(parameters.revInterval());
        }
        assertThat(parameters.getIntervalOffset(), is(-23 * 60 + 30));

        assertFalse(parameters.revInterval());
        assertThat(parameters.getIntervalOffset(), is(-23 * 60 + 30));
    }

    @Test
    public void testFastforwardInterval() {
        parameters.setOffsetIncrement(15);
        parameters.revInterval();

        assertTrue(parameters.ffwdInterval());
        assertThat(parameters.getIntervalOffset(), is(0));

        assertFalse(parameters.ffwdInterval());
        assertThat(parameters.getIntervalOffset(), is(0));
    }

    @Test
    public void testGoRealtime() {
        parameters.setOffsetIncrement(15);

        assertFalse(parameters.goRealtime());

        parameters.revInterval();

        assertTrue(parameters.goRealtime());
        assertThat(parameters.getIntervalOffset(), is(0));
    }

    @Test
    public void testGetIntervalDuration() {
        assertThat(parameters.getIntervalDuration(), is(60));

        assertTrue(parameters.setIntervalDuration(120));
        assertThat(parameters.getIntervalDuration(), is(120));

        assertFalse(parameters.setIntervalDuration(120));
        assertThat(parameters.getIntervalDuration(), is(120));
    }


}

File: app/src/test/java/org/blitzortung/android/data/ParametersTest.kt
package org.blitzortung.android.data

import org.assertj.core.api.Assertions.assertThat
import org.junit.Test


class ParametersTest {

    private val parameters = Parameters()
    private val history = History()

    @Test
    fun returnsDuration() {
        assertThat(parameters.intervalDuration).isEqualTo(60)
    }

    @Test
    fun returnsOffset() {
        assertThat(parameters.intervalOffset).isEqualTo(0)
    }

    @Test
    fun isRealtime() {
        assertThat(parameters.isRealtime()).isTrue()
    }

    @Test
    fun returnsNoRealtime() {
        val result = parameters.animationStep(history)

        assertThat(result.isRealtime()).isFalse()
    }

    @Test
    fun goRealtime() {
        val result = parameters.animationStep(history).goRealtime()

        assertThat(result.isRealtime()).isTrue()
    }


    @Test
    fun goToHistory() {
        val result = parameters.animationStep(history)
        assertThat(result.isRealtime()).isFalse()
        assertThat(parameters.intervalOffset).isEqualTo(0)
    }

    @Test
    fun defaultSliderEndPosition() {
        assertThat(parameters.intervalPosition(history)).isEqualTo(46)
        assertThat(parameters.intervalMaxPosition(history)).isEqualTo(46)
    }

    @Test
    fun sliderEndPositionForUnlimitedHistory() {
        val unlimitedHistory = History(5, 120, false)

        assertThat(parameters.intervalPosition(unlimitedHistory)).isEqualTo(24)
        assertThat(parameters.intervalMaxPosition(unlimitedHistory)).isEqualTo(24)
    }

    @Test
    fun sliderStartPosition() {
        val result = parameters.animationStep(history)

        assertThat(result.intervalPosition(history)).isEqualTo(0)
        assertThat(result.intervalMaxPosition(history)).isEqualTo(46)
    }

    @Test
    fun withPositionAtStart() {
        val result = parameters.withPosition(0, history)

        assertThat(result.intervalOffset).isEqualTo(-1380)
    }

    @Test
    fun withPositionAtEnd() {
        val result = parameters.withPosition(46, history)

        assertThat(result.intervalOffset).isEqualTo(0)
    }

    @Test
    fun withIntervalDurationUpdate() {
        val result = parameters.withIntervalDuration(120)

        assertThat(result.intervalMaxPosition(history)).isEqualTo(44)
    }

    @Test
    fun withLocalReference() {
        val localReference = LocalReference(5, 6, dataArea = 5)
        val result = parameters.copy(localReference = localReference)

        assertThat(result.localReference).isEqualTo(localReference)
    }
}
